src/core/server/saved_objects/migrations/core/document_migrator.ts

```diff
@@ -754,7 +754,7 @@ function migrateProp(
 
   for (const { version, transform, transformType } of applicableTransforms(migrations, doc, prop)) {
     const currentVersion = propVersion(doc, prop);
-    if (currentVersion && Semver.gt(currentVersion, version)) {
+    if (currentVersion && Semver.gt(currentVersion, version) && transformType !== 'reference') {
       // the previous transform function increased the object's migrationVersion; break out of the loop
       break;
     }
```

Comment:
This fixes the bug at hand, but maybe we want to change more than this.

I mentioned in the linked issue and I will mention again here, we had a conversation about this last year: https://github.com/elastic/kibana/pull/80945#discussion_r535928706

The comment chain on that PR is very long, but here's a screenshot of the discussion in question:

![image](https://user-images.githubusercontent.com/5295965/144291907-d2a52cce-3992-4096-a60b-0119cbb4e2d2.png)

And here is the commit that my comment linked to: e27e6be8e0f1078135774b1dfe572882a2a53e5a

---

SO: 

1. Technically you can define a "migration" transform that skips your object's "conversion" transform. It's not a problem that we realistically have to worry about, but maybe we want to prevent that from happening?
2. While we're at it, maybe we want to reconsider this whole idea of allowing "leapfrog" migrations. AFAIK it was supported in the original DocumentMigrator implementation but it has never been used by consumers. We already had unit tests for it, true, but in hindsight it seems to add unnecessary complexity for a feature that no consumers are using (that we are aware of).

So, maybe we should just rip out these 3 lines of code and modify the tests accordingly. WDYT?

Reply:
I agree, this code is really hard to follow so any complexity we can remove would be worth it.

Reply:
OK, I will make additional changes tomorrow to that effect

Reply:
Changed in 060dff5348324368f4d1cde6bcd80289a1b48cf5.

I noticed some comments at the top of the document that mention that migration functions can increase the `migrationVersion`, but cannot decrease it. I changed those comments accordingly and added an explicit disclaimer in our dev docs that changing `migrationVersion`, `type`, and `id` fields is **not** supported.

I noticed there is already an `assertNoDowngrades` validation function that checks to see if a consumer has decreased the `migrationVersion` and throws an error: https://github.com/elastic/kibana/blob/060dff5348324368f4d1cde6bcd80289a1b48cf5/src/core/server/saved_objects/migrations/core/document_migrator.ts#L812-L834

This is good to keep, because decreasing the `migrationVersion` could cause an infinite loop.

I didn't **add** validation to check for an _increased_ `migrationVersion`, my reasoning is:
* The impact is low -- a migration function that increases the `migrationVersion` simply won't work as intended, e.g., it won't "leapfrog" other transforms
* I want to reduce complexity and testing burden
* We never explicitly stated in our dev docs that we supported increasing the `migrationVersion`, and now we are explicitly stating that we _don't_ support it

But if you think I should add validation anyway, I'll do it, just let me know.